src/main/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -856,7 +856,9 @@ private void markDuplicatePairs(final List<ReadEndsForMarkDuplicates> list) {
 
                 if (end.isOpticalDuplicate && this.opticalDuplicateIndexes != null) {
                     this.opticalDuplicateIndexes.add(end.read1IndexInFile);
-                    this.opticalDuplicateIndexes.add(end.read2IndexInFile);
+                    if(end.read2IndexInFile != end.read1IndexInFile) {
```

Comment:
```suggestion
                    if (end.read2IndexInFile != end.read1IndexInFile) {
```

Reply:
also, it's quite mysterious (without more explanation) as to why you only need to do this if the index is different...would it be possible to refactor this and the code that pulls from the iterator, so that they are closer to each other and then add comments that make it more clear?